contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,198 @@
+
+After defining the function, we call it, and we call another function `.then()` upon the result. `.then()` is used to consume the result from a fulfilled Promise and act on it. In this case, we are using it to log our result. More importantly, it lets us chain promises to avoid the deep nesting known as callback hell.
+
+Here's the callback hell code rewritten with Promises.
+
+```javascript
+
+function update(numTimes) {
+    return new Promise((resolve, reject) => {
+        var x = "I have not been updated";
+        setTimeout(function() {
+            x = "I have been updated " + numTimes + " times.";
+            console.log(x);
+            resolve(numTimes);
+        }, 1000);
+    });
+}
+
+update(1).then((result) => update(result + 1))
+.then((result) => update(result + 1));
+
+```
+
+Now we update our function to take in a parameter `numTimes` so that we can log how many times `x` was updated. `then()` will return a Promise, which you can then again call `.then()`. By doing this, we can chain our Promises and ensure that execution of the code goes in the order that we require. In addition, it also allows your variables to be correctly passed down, ensuring that you will not have to deal with any undefined variables along the way.
+
+Using Promises helps your code to be clean and easy to read, which makes it easy for anyone to instantly know what the code does, instead of having to trace through the code written with callbacks.
+
+## How Promises are used
+
+There are many interesting ways to use Promises. Here we cover HTTP requests in Javascript, and Disk I/O in Node.js. No prior knowledge of Node.js is necessary.
+
+### HTTP Requests
+
+Earlier in this chapter, HTTP requests were mentioned as an example of an asynchronous operation. By using Promises, you would be able to act on the result from the request without having to use callbacks or wait an arbitrarily set amount of time for the response to be returned.
+
+The following code sends a GET request to a URL and logs the body of the response using Promises. By using Promises instead of callbacks, we have clean code and improved performance as the code is able to run in the background.
+
+Code snippet from [Promise MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
+
+```javascript
+
+function myAsyncFunction(url) {
+  return new Promise((resolve, reject) => {
+    const xhr = new XMLHttpRequest();
+    xhr.open("GET", url);
+    xhr.onload = () => resolve(xhr.responseText);
+    xhr.onerror = () => reject(xhr.statusText);
+    xhr.send();
+  });
+}
```

Comment:
May be show how this is used too, to be consistent with other examples you have given?

Reply:
Added a snippet to show how the function is called and handled.